src/profiling/with_perf.md

```diff
@@ -0,0 +1,299 @@
+# Profiling with perf
+
+This is a guide for how to profile rustc with [perf](https://perf.wiki.kernel.org/index.php/Main_Page).
+
+## Initial steps
+
+- Get a clean checkout of rust-lang/master, or whatever it is you want
+  to profile.
+- Set the following settings in your `config.toml`:
+  - `debuginfo-lines = true`
+  - `use-jemalloc = false` — lets you do memory use profiling with valgrind 
+  - leave everything else the defaults
+- Run `./x.py build` to get a full build
+- Make a rustup toolchain (let's call it `rust-prof`) pointing to that result
+  - `rustup toolchain link <path-to-toolchain>`
+  
+## Gathering a perf profile
+
+perf is an excellent tool on linux that can be used to gather and
+analyze all kinds of information. Mostly it is used to figure out
+where a program spends its time. It can also be used for other sorts
+of events, though, like cache misses and so forth.
+
+### The basics
+
+The basic `perf` command is this:
+
+```
+perf record -F99 --call-graph dwarf XXX
+```
+
+The `-F99` tells perf to sample at 99 Hz, which avoids generating too
```

Comment:
Everything about doing the `perf` recording -- from here to the end of the "Gathering NLL data" section -- is doing things the hard way! `rustc-perf` is able to do all this stuff for you, specifically the [`collector profile perf-record`](https://github.com/rust-lang-nursery/rustc-perf/blob/master/collector/README.md#profiling-local-builds) command.

It already uses `-F99` and `--call-graph dwarf`, it does dependency stuff for you and only measures the compilation of the final crate, you don't need to distinguish between library and non-library crates, you can specify what kind of "run" you want (Clean, Nll, BaseIncr, CleanIncr, PatchedIncrs, All, or any combination of the above), and also what kind of "build" (Check, Debug, Opt, All, or a combination).

Reply:
Ah, good to know.

Reply:
Learn something new =)